wire/netaddress.go

```diff
@@ -90,12 +83,13 @@ func NewNetAddress(addr *net.TCPAddr, services ServiceFlag) *NetAddress {
 func readNetAddress(r io.Reader, pver uint32, na *NetAddress, ts bool) error {
 	var services ServiceFlag
 	var ip [16]byte
+	timestamp := na.Timestamp
```

Comment:
Yes, it is allocated on the heap (as can be seen in the profile I pasted).  Interfaces in Go (of which the readers all are) always force the contained thing to be boxed.

In the case of an `addr` message, there is a slice of `NetAddress` which is allocated as a single contiguous block on the heap.  This function then reads directly into those structs that are already on the heap and thus doesn't need another allocation.  If you, however, introduce a local stack variable as is being done here and read into it, the read forces it to the heap because of the boxing.